### PR TITLE
Update dhcpmon rx/tx packet filtering and fix server rx count

### PR DESCRIPTION
**Pull request by kellyyeh**
#### Why I did it
Dhcpmon had incorrect RX count for server side packets. It does not raise any false alarms, but could miss catching server side packet count mismatch between snapshot and current counter.

Add debug mode which prints counter to syslog 

#### How I did it
Due to dualtor inbound filter requirement, there are currently two filters, each for listening to rx / tx packets.
Originally, we opened up an rx/tx socket for each interface specified, which causes duplicate socket. Now we initialize the sockets only once. Both sockets are not binded to an interface, and we use vlan to interface mapping to filter packets. For inbound uplinks, we use a portchannel to interface mapping.

Previous dhcpmon counter before dual tor change:
[    Agg-Vlan1000- Current rx/tx] Discover:         1/        4, Offer:         1/        1, Request:         3/       12, ACK:         1/        1
[            eth0- Current rx/tx] Discover:         0/        0, Offer:         0/        0, Request:         0/        0, ACK:         0/        0
[            eth0- Current rx/tx] Discover:         0/        0, Offer:         0/        0, Request:         0/        0, ACK:         0/        0
[  PortChannel104- Current rx/tx] Discover:         0/        1, Offer:         0/        0, Request:         0/        3, ACK:         0/        0
[  PortChannel103- Current rx/tx] Discover:         0/        1, Offer:         0/        0, Request:         0/        3, ACK:         0/        0
[  PortChannel102- Current rx/tx] Discover:         0/        2, **Offer:         1/        0,** Request:         0/        6, **ACK:         1/        0**
[  PortChannel101- Current rx/tx] Discover:         0/        0, Offer:         0/        0, Request:         0/        0, ACK:         0/        0
[        Vlan1000- Current rx/tx] Discover:         1/        0, Offer:         0/        1, Request:         3/        0, ACK:         0/        1
[    Agg-Vlan1000- Current rx/tx] Discover:         1/        4, Offer:         1/        1, Request:         3/       12, ACK:         1/        1

Dhcpmon counter after this PR:
[  PortChannel104- Current rx/tx] Discover:         0/        1, **Offer:         0/        0**, Request:         0/        3, **ACK:         0/        0**
[  PortChannel103- Current rx/tx] Discover:         0/        1, **Offer:         0/        0**, Request:         0/        3, **ACK:         0/        0**
[  PortChannel102- Current rx/tx] Discover:         0/        2, **Offer:         1/        0**, Request:         0/        6, **ACK:         1/        0**
[  PortChannel101- Current rx/tx] Discover:         0/        0, **Offer:         0/        0**, Request:         0/        0, **ACK:         0/        0**
[        Vlan1000- Current rx/tx] Discover:         1/        0, Offer:         0/        1, Request:         3/        0, ACK:         0/        1
[    Agg-Vlan1000- Current rx/tx] Discover:         1/        4, Offer:         1/        1, Request:         3/       12, ACK:         1/        1

#### How to verify it
Ran dhcp relay test to send all four packets in singles and batches on both single ToR and dual ToR. Counter was as expected.

Standby tor:
[  PortChannel104- Current rx/tx] Discover:         0/        0, Offer:         0/        0, Request:         0/        0, ACK:         0/        0
[  PortChannel103- Current rx/tx] Discover:         0/        0, Offer:         0/        0, Request:         0/        0, ACK:         0/        0
[  PortChannel102- Current rx/tx] Discover:         0/        0, Offer:         0/        0, Request:         0/        0, ACK:         0/        0
[  PortChannel101- Current rx/tx] Discover:         0/        0, Offer:         0/        0, Request:         0/        0, ACK:         0/        0
[        Vlan1000- Current rx/tx] Discover:         1/        0, Offer:         0/        0, Request:         1/        0, ACK:         0/        0
[    Agg-Vlan1000- Current rx/tx] Discover:         1/        0, Offer:         0/        0, Request:         1/        0, ACK:         0/        0

Active tor:
[  PortChannel104- Current rx/tx] Discover:         0/       11, Offer:         0/        0, Request:         0/       11, ACK:         0/        0
[  PortChannel103- Current rx/tx] Discover:         0/       14, Offer:         0/        0, Request:         0/       14, ACK:         0/        0
[  PortChannel102- Current rx/tx] Discover:         0/       10, Offer:         1/        0, Request:         0/       10, ACK:         1/        0
[  PortChannel101- Current rx/tx] Discover:         0/       13, Offer:         0/        0, Request:         0/       13, ACK:         0/        0
[        Vlan1000- Current rx/tx] Discover:         1/        0, Offer:         0/        1, Request:         1/        0, ACK:         0/        1
[    Agg-Vlan1000- Current rx/tx] Discover:         1/       48, Offer:         1/        1, Request:         1/       48, ACK:         1/        1
#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [x] 202012
- [ ] 202106
- [ ] 202111
- [x] 202205
- [x] 202211

### CODE REVIEW

**Comment by jcaiMR** [comment]
A little bit confused about the testing result:

Dhcpmon counter after this PR:
[ PortChannel104- Current rx/tx] Discover: 0/ 1, Offer: 1/ 0, Request: 0/ 3, ACK: 1/ 0
[ PortChannel103- Current rx/tx] Discover: 0/ 1, Offer: 1/ 0, Request: 0/ 3, ACK: 1/ 0
[ PortChannel102- Current rx/tx] Discover: 0/ 2, Offer: 1/ 0, Request: 0/ 6, ACK: 1/ 0
[ PortChannel101- Current rx/tx] Discover: 0/ 0, Offer: 1/ 0, Request: 0/ 0, ACK: 1/ 0
[ Vlan1000- Current rx/tx] Discover: 1/ 0, Offer: 1/ 1, Request: 3/ 0, ACK: 1/ 1
[ Agg-Vlan1000- Current rx/tx] Discover: 1/ 4, Offer: 1/ 1, Request: 3/ 12, ACK: 1/ 1

Here 1 Discover from client, with 4 help address, 4 Discover relayed to  DHCP server, due to ECMP port channel tx counters = 1+1+2+0, which is right.
Then 4 servers give 4 Offer and send to portchannel interfaces, Portchannel counter = 1+1+1+1
Then isc-dhcp will send 4 offer to client, suppose Vlan1000 Offer RX should be 4 and Offer TX should always same as RX
But here the counter is 1/1 which seems not right. 
Same as ACK counter.

**Review by jcaiMR — COMMENTED** [review]
A little bit confused about the testing result:

Dhcpmon counter after this PR:
[ PortChannel104- Current rx/tx] Discover: 0/ 1, Offer: 1/ 0, Request: 0/ 3, ACK: 1/ 0
[ PortChannel103- Current rx/tx] Discover: 0/ 1, Offer: 1/ 0, Request: 0/ 3, ACK: 1/ 0
[ PortChannel102- Current rx/tx] Discover: 0/ 2, Offer: 1/ 0, Request: 0/ 6, ACK: 1/ 0
[ PortChannel101- Current rx/tx] Discover: 0/ 0, Offer: 1/ 0, Request: 0/ 0, ACK: 1/ 0
[ Vlan1000- Current rx/tx] Discover: 1/ 0, Offer: 1/ 1, Request: 3/ 0, ACK: 1/ 1
[ Agg-Vlan1000- Current rx/tx] Discover: 1/ 4, Offer: 1/ 1, Request: 3/ 12, ACK: 1/ 1

Here 1 Discover from client, with 4 help address, 4 Discover relayed to DHCP server, due to ECMP port channel tx counters = 1+1+2+0, which is right.
Then 4 servers give 4 Offer and send to portchannel interfaces, Portchannel Offer RX counter = 1+1+1+1
Then isc-dhcp will send 4 offers to client, suppose Vlan1000 Offer RX should be 4 and Vlan Offer TX should always same as Vlan Offer RX which is also 4.
But here the counter is 1/1 which seems not right. Same as ACK.

Another thing, is this testing under a real client DHCP online procedural ? The Request counter seems not match with ACK counter.

File: src/dhcpmon/src/dhcp_device.cpp [hunk 1]
```diff
@@ -61,36 +62,37 @@ std::shared_ptr<swss::Table> mStateDbMuxTablePtr = std::make_shared<swss::Table>
...
 std::unordered_map<std::string, std::string> vlan_map;
 
+/* list of portchannels */
+std::vector<std::string> portchannel_list;
+
 /** Berkeley Packet Filter program for "udp and (port 67 or port 68)".
  * This program is obtained using the following command tcpdump:
- * `tcpdump -dd "outbound and udp and (port 67 or port 68)"`
+ * `tcpdump -dd "udp and (port 67 or port 68)"`
  */
```
Comment (jcaiMR) [inline src/dhcpmon/src/dhcp_device.cpp:71]: Hi Kelly, can we still use origin bpf filter, and follow below logic ?

Inbound listen all interfaces
1. int is vlan member, update Ethernet Discover/Request RX
2. int is vlan, update Vlan Discover/Request RX
3. int is port-channel, update port-channel Offer/Ack RX

Outbond  listen all interfaces
1. int is vlan member, update Ethernet Offer/Ack TX
2. int is vlan, update Vlan Offer/Ack Tx
3. int is port-channel. update port-channel Discover/Request TX

Vlan AGG Discover RX = Vlan Discover RX, TX = port-channel Discover TX
Vlan AGG Offer RX = port-channel Offer RX, TX = Ethernet Offer TX
Vlan AGG Request RX = Vlan Request RX, TX = port-channel Request TX
Vlan AGG Ack RX = port-channel Ack RX, TX = Ethernet Ack TX

Comment (kellyyeh) [reply src/dhcpmon/src/dhcp_device.cpp:71]: Hi Jie, what you proposed is the logic we have right now before this PR(with a bug in offer/ack rx), but the agg counter is different from the original dhcpmon counter.

Original:
[ eth0- Current rx/tx] Discover: 0/ 0, Offer: 0/ 0, Request: 0/ 0, ACK: 0/ 0
[ eth0- Current rx/tx] Discover: 0/ 0, Offer: 0/ 0, Request: 0/ 0, ACK: 0/ 0
[ PortChannel104- Current rx/tx] Discover: 0/ 1, Offer: 0/ 0, Request: 0/ 3, ACK: 0/ 0
[ PortChannel103- Current rx/tx] Discover: 0/ 1, Offer: 0/ 0, Request: 0/ 3, ACK: 0/ 0
[ PortChannel102- Current rx/tx] Discover: 0/ 2, Offer: 1/ 0, Request: 0/ 6, ACK: 1/ 0
[ PortChannel101- Current rx/tx] Discover: 0/ 0, Offer: 0/ 0, Request: 0/ 0, ACK: 0/ 0
[ Vlan1000- Current rx/tx] Discover: 1/ 0, Offer: 0/ 1, Request: 3/ 0, ACK: 0/ 1
[ Agg-Vlan1000- Current rx/tx] Discover: 1/ 4, **Offer: 1/ 1,** Request: 3/ 12, **ACK: 1/ 1**

Current:
[ eth0- Current rx/tx] Discover: 0/ 0, Offer: 0/ 0, Request: 0/ 0, ACK: 0/ 0
[ eth0- Current rx/tx] Discover: 0/ 0, Offer: 0/ 0, Request: 0/ 0, ACK: 0/ 0
[ PortChannel104- Current rx/tx] Discover: 0/ 1, Offer: 1/ 0, Request: 0/ 3, ACK: 1/ 0
[ PortChannel103- Current rx/tx] Discover: 0/ 1, Offer: 1/ 0, Request: 0/ 3, ACK: 1/ 0
[ PortChannel102- Current rx/tx] Discover: 0/ 2, Offer: 1/ 0, Request: 0/ 6, ACK: 1/ 0
[ PortChannel101- Current rx/tx] Discover: 0/ 0, Offer: 1/ 0, Request: 0/ 0, ACK: 1/ 0
[ Vlan1000- Current rx/tx] Discover: 1/ 0, Offer: 0/ 1, Request: 3/ 0, ACK: 0/ 1
[ Agg-Vlan1000- Current rx/tx] Discover: 1/ 4, **Offer: 4/ 1,** Request: 3/ 12, **ACK: 4/ 1**

I wanted to keep the agg counter consistent by changing:
Vlan AGG Offer/Ack RX = port-channel Offer/Ack RX
to:
Vlan AGG Offer/Ack RX = Vlan Offer/Ack RX
And keeping track of Offer/Ack RX on Portchannels but not adding them to agg-counter.

Though according to the this function of checking if dhcp relay message is healthy here https://github.com/sonic-net/sonic-buildimage/blob/f8a393c3a176d120bc6472f3b56936ef85916d4a/src/dhcpmon/src/dhcp_device.cpp#L368, the new logic would work just fine.
There's no official documentation of dhcpmon counter. We could potentially change it as we like as long as the disparity alarm works fine

**Comment by kellyyeh** [comment]
@jcaiMR 

> Then isc-dhcp will send 4 offers to client, suppose Vlan1000 Offer RX should be 4 and Vlan Offer TX should always same as Vlan Offer RX which is also 4.
But here the counter is 1/1 which seems not right. Same as ACK.

This is the original dhcpmon counter count as mentioned in the below reply. 

> Then isc-dhcp will send 4 offers to client, suppose Vlan1000 Offer RX should be 4 and Vlan Offer TX should always same as Vlan Offer RX which is also 4.
But here the counter is 1/1 which seems not right. Same as ACK.

This is the counter after running dhcp_relay default test from sonic-mgmt. I ran it with the original dhcpmon and dhcpmon after dualtor change. It sends one packet of each, and the request count is the total count of dhclient receiving ack and sending the response and the test sending a mock request packet. Not sure why Ack does not match.

**Review by jcaiMR — APPROVED** [review]

**Review by yxieca — APPROVED** [review]

**Pull request merged**